clang/lib/AST/ASTContext.cpp

```diff
@@ -11730,6 +11730,55 @@ bool ASTContext::containsCapabilities(QualType Ty) const {
   return Ret;
 }
 
+bool ASTContext::cannotContainCapabilities(const RecordDecl *RD) const {
+  for (auto i = RD->field_begin(), e = RD->field_end(); i != e; ++i) {
+    const QualType Ty = i->getType();
+    if (Ty->isCHERICapabilityType(*this))
+      return false;
+    else if (const RecordType *RT = Ty->getAs<RecordType>()) {
+      if (!cannotContainCapabilities(RT->getDecl()))
```

Comment:
The double negatives make this function implementation rather ugly, but I think it makes more sense for callers such as CodeGenTypes::copyShouldPreserveTagsForPointee().